internal/pkg/agent/protection/policy.go

```
@@ -0,0 +1,196 @@
+//
+// There could be few cases this function needs to handle:
+// 1. The old policy format that doesn't have signature or protected data. Policy should returned without modifications or errors.
+// 2. New policy where the signatureValidationKey is nil. No previously known signature validation key, try to read one from the policy, validate and return.
+// 3. New policy format with the signatureValidationKey. Validate the signed data normally.
+func ValidatePolicySignature(log *logger.Logger, policy map[string]interface{}, signatureValidationKey []byte) (map[string]interface{}, []byte, error) {
+	var err error
+
+	const logPrefix = "Validate policy signature:"
+
+	log.Infof("%v passed signature validation key length: %v", logPrefix, len(signatureValidationKey))
+
+	// Try to get the signature validation key from the policy itself if there is no previously known signature validation key
+	if len(signatureValidationKey) == 0 {
+		signatureValidationKey, err = getPolicySignatureValidationKey(policy)
+		log.Infof("%v policy signature validation key length: %v, err: %v", logPrefix, len(signatureValidationKey), err)
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return nil, nil, err
+		}
+	}
+
+	// Validate the signature
+	if len(signatureValidationKey) != 0 {
+		// Read protected data and signature from the policy
+		data, signature, err := getPolicySignedDataAndSignature(policy)
+		log.Infof("%v policy data length: %v, signature length: %v, err: %v", logPrefix, len(data), len(signature), err)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Validate signed data
+		err = ValidateSignature(data, signature, signatureValidationKey)
+		log.Infof("%v policy signature validation result: %v", logPrefix, err)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// Unmarshal signed layer of the policy
+		var signedLayer map[string]interface{}
+		if json.Valid(data) {
```

If this is false should you return an error?